File: infra/gateway/src/gateway/auth_guard.py
# SPDX-License-Identifier: Apache-2.0
"""Boot-time guard on the gateway's authentication escape hatch (CTO-268, gateway half).

WHY this file exists. ``TALLY_REQUIRE_API_KEY=false`` is one environment variable that turns
authentication OFF for the whole gateway: ``/v1/batches`` accepts whatever ``tenant_id`` the caller
puts in the body, and the control-plane service-token gate on ``/v1/tenant/*`` goes quiet too,
because that gate is active only when auth is ON and a token is configured (see
``app._require_service_token``). Anyone who can reach the port can then write spans against any
tenant and read and rewrite any tenant's control-plane configuration. That is correct and wanted for
``make up``, for the test suite and for a laptop; it is a catastrophe on a reachable deployment.

The web tier got this guard first (``web/lib/authGuard.ts``, PR #345) and its author deferred the
gateway half for a stated reason: ``gateway/config.py`` had no notion of a deployment environment,
so there was nothing to test ``TALLY_REQUIRE_API_KEY`` against, and introducing that notion is a
design decision rather than a line of code. This module makes that decision.

THE DECISION, and the honest limitation in it. The web tier gets ``NODE_ENV=production`` for free:
``next start`` and the standalone server set it, so the guard fires without an operator doing
anything. A Python process has no equivalent signal, and there is nothing in the gateway's existing
configuration that reliably means "this is a real deployment" (a Postgres DSN pointing at a remote
host is a guess, and guessing is exactly what the honest-under-uncertainty invariant forbids). So
this introduces an explicit one, ``TALLY_ENV``, defaulting to ``development``:

* ``development`` / ``dev`` / ``local`` / ``test``: no guard. ``make up``, CI and pytest are
  unaffected and need no new variable.
* ``production`` / ``prod`` / ``staging`` / ``stage``: auth must be on, OR the operator must say
  ``TALLY_ALLOW_INSECURE_NO_AUTH=1`` and live with a warning banner on every boot.
* anything else: treated as a deployment, because an unrecognized value is far more likely to be a
  typo in a deployment manifest than a laptop, and the fail-closed direction is the safe one.

The limitation is that an operator who never sets ``TALLY_ENV`` is not protected. That is real, and
it is why the deployment manifests must set it (``deploy/aws/README.md`` says so, and the ECS task
definition already sets ``TALLY_REQUIRE_API_KEY=true`` so the guard is a backstop there, not the
thing that saves it). A guard that defaulted the other way would refuse to start every developer's
``make up`` and every CI run, which is how guards get deleted.

WHY THIS DOES NOT BREAK THE AUTH-DISABLED SHAPE. ``infra/edge-proxy/README.md`` documents
``EDGE_PROXY_TENANT_ID`` as "Required when the gateway runs with auth disabled", so an auth-disabled
gateway is a supported configuration somewhere and must keep working. It does: it is untouched in
every development environment, and it remains available in a deployment behind the same explicit
opt-in the web tier introduced. We deliberately reuse ``TALLY_ALLOW_INSECURE_NO_AUTH`` rather than
inventing a second spelling, so one operator sentence covers both tiers of one deployment.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass

logger = logging.getLogger(__name__)

#: Env keys this guard reads. Named once so the message and the checks cannot drift.
ENV_VAR = "TALLY_ENV"
REQUIRE_API_KEY_ENV = "TALLY_REQUIRE_API_KEY"
ALLOW_INSECURE_ENV = "TALLY_ALLOW_INSECURE_NO_AUTH"

#: Environment names that mean "a laptop, CI, or the test suite": the guard stands down.
LOCAL_ENVIRONMENTS: frozenset[str] = frozenset({"development", "dev", "local", "test", "ci"})

#: Environment names that mean "a real, reachable deployment". Anything unrecognized joins them.
DEPLOYED_ENVIRONMENTS: frozenset[str] = frozenset({"production", "prod", "staging", "stage"})


class InsecureAuthConfigError(RuntimeError):
    """Auth is off in a deployed environment with no explicit opt-in. Raised at boot."""


def is_truthy(value: object) -> bool:
    """Same truthiness spelling ``web/lib/authGuard.ts`` and ``deploy/demo/lib-tenant.sh`` accept."""
    if isinstance(value, bool):
        return value
    if not isinstance(value, str):
        return False
    return value.strip().lower() in ("1", "true", "yes", "on")


def is_deployed_environment(env: object) -> bool:
    """Whether ``TALLY_ENV`` names a real deployment. Unrecognized values fail closed (deployed)."""
    name = (env if isinstance(env, str) else "").strip().lower()
    if not name:
        # Empty is the default and means "unset", which is a laptop until someone says otherwise.
        return False
    return name not in LOCAL_ENVIRONMENTS


@dataclass(frozen=True, slots=True)
class AuthGuardVerdict:
    """What the guard decided. ``kind`` is one of ``ok`` / ``insecure-allowed`` / ``refuse``."""

    kind: str
    message: str = ""


def _refusal_message(env_name: str) -> str:
    return f"""
================================================================================
ai-tally gateway REFUSES TO START: authentication is disabled in {ENV_VAR}={env_name}.

{REQUIRE_API_KEY_ENV} is false, which turns authentication OFF for the whole
gateway, not just for ingest:

  * POST /v1/batches trusts the tenant_id in the request body, so any caller can
    write spans, cost and business events against ANY tenant.
  * The control-plane service-token gate on /v1/tenant/* is inert (it is enforced
    only when auth is on), so any caller can read and rewrite any tenant's
    connectors, budgets, guardrails and ingest keys.

There is no safe fallback here, so this process stops instead of serving.

You are in ONE OF TWO situations:

  1. You want a REAL deployment. This is the normal case.

         {REQUIRE_API_KEY_ENV}=true
         TALLY_GATEWAY_SERVICE_TOKEN=<a long random server-only secret>

     The service token authenticates the WEB SERVER on /v1/tenant/* calls; the
     gateway already refuses to boot with auth on and that token empty.
     deploy/aws/ecs/gateway.taskdef.json sets both.

  2. You genuinely want an OPEN gateway, the way a laptop or the public demo runs.

     Then say so, explicitly:

         {ALLOW_INSECURE_ENV}=1

     Only do this on a port nobody else can reach, or behind access control you
     supply yourself. Note that the edge proxy's EDGE_PROXY_TENANT_ID exists for
     exactly this shape (infra/edge-proxy/README.md), and it keeps working.

     Or set {ENV_VAR}=development, which is what a laptop is.

Checked: {ENV_VAR}={env_name}, {REQUIRE_API_KEY_ENV}=false, {ALLOW_INSECURE_ENV} not set.
================================================================================
""".strip()


def _insecure_warning(env_name: str) -> str:
    return f"""
================================================================================
ai-tally gateway WARNING: serving with NO AUTHENTICATION.

{ENV_VAR}={env_name} and {REQUIRE_API_KEY_ENV} is false, with
{ALLOW_INSECURE_ENV} set, so this gateway starts wide open: ingest trusts the
tenant_id in the body and the control plane is ungated.

This is a deliberate configuration. If you did not mean it, set
{REQUIRE_API_KEY_ENV}=true and give the gateway a service token.
================================================================================
""".strip()


def check_auth_config(
    *, env: object, require_api_key: object, allow_insecure: object
) -> AuthGuardVerdict:
    """Decide whether this process may serve. Pure, so the decision table is testable.

    ``assert_auth_config`` is the one that stops the boot; this is separated out so tests and the
    allowed-but-insecure log path can inspect the decision without catching an exception.
    """
    if require_api_key is True or is_truthy(require_api_key):
        return AuthGuardVerdict("ok")
    if not is_deployed_environment(env):
        return AuthGuardVerdict("ok")
    env_name = (env if isinstance(env, str) else "").strip()
    if is_truthy(allow_insecure):
        return AuthGuardVerdict("insecure-allowed", _insecure_warning(env_name))
    return AuthGuardVerdict("refuse", _refusal_message(env_name))


def assert_auth_config(settings: object) -> AuthGuardVerdict:
    """Raise :class:`InsecureAuthConfigError` on a refusal, warn loudly on an opted-in open gateway.

    Called from the FastAPI lifespan, which is the gateway's equivalent of the web tier's
    ``instrumentation.ts``: it runs once, before the first request is served, in every way this
    process is started (uvicorn, the container entrypoint, a TestClient). A per-request check would
    be skippable by any route that forgot it.
    """
    verdict = check_auth_config(
        env=getattr(settings, "env", ""),
        require_api_key=getattr(settings, "require_api_key", False),
        allow_insecure=getattr(settings, "allow_insecure_no_auth", ""),
    )
    if verdict.kind == "refuse":
        raise InsecureAuthConfigError(verdict.message)
    if verdict.kind == "insecure-allowed":
        logger.warning("%s", verdict.message)
    return verdict

File: web/lib/authGuard.ts
// SPDX-License-Identifier: Apache-2.0
// Boot-time guard on the dev escape hatch (CTO-268).
//
// WHY this file exists. `TALLY_DEV_TENANT` is one environment variable that turns authentication
// OFF everywhere at once: `middleware.ts` exports a pass-through instead of `clerkMiddleware()`,
// `app/layout.tsx` mounts no `ClerkProvider`, `getTenant()` short-circuits to the pinned tenant
// instead of resolving the signed-in Clerk org, and `canManage()` returns true for every visitor,
// so key mint / rotate / revoke is ungated too. That is correct and wanted for `make up`, for CI,
// for the vitest suite and for the public demo VM. It is a catastrophe on a real instance: anyone
// with the URL reads every number in the system.
//
// The trap is concrete. `deploy/demo/` is the only genuine one-command deploy in the repo, so it is
// the thing an operator copies when standing up a real instance, and it exports `TALLY_DEV_TENANT`
// on purpose. The production manifests set it empty today, but nothing enforced that they stay that
// way. This module is that enforcement.
//
// THE CONTRACT. Disabling auth in a production build now takes TWO variables, the second of which
// nobody sets by accident:
//
//   TALLY_DEV_TENANT              pins the tenant (unchanged)
//   TALLY_ALLOW_INSECURE_NO_AUTH  "yes, I mean it, serve this with no authentication"
//
// One variable alone in a production build is a hard boot failure. Two is allowed and prints a
// standing warning on every boot, so the demo keeps working but stops being silent.
//
// Pure and dependency-free on purpose: it is imported from `instrumentation.ts` (Node runtime),
// from `middleware.ts` (Edge runtime) and from the vitest suite, so it must not import
// `server-only`, Clerk, or anything Node-specific.

/** Env keys this guard reads. Named once so the message and the checks cannot drift. */
export const DEV_TENANT_ENV = "TALLY_DEV_TENANT";
export const ALLOW_INSECURE_ENV = "TALLY_ALLOW_INSECURE_NO_AUTH";

/** The subset of `process.env` the guard needs. Passed explicitly so tests never mutate globals. */
export interface AuthGuardEnv {
  NODE_ENV?: string;
  NEXT_PHASE?: string;
  TALLY_DEV_TENANT?: string;
  TALLY_ALLOW_INSECURE_NO_AUTH?: string;
}

/** Raised when the escape hatch is active in a production build with no explicit opt-in. */
export class InsecureAuthConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InsecureAuthConfigError";
  }
}

/** Same truthiness spelling `deploy/demo/lib-tenant.sh` accepts, so shell and app agree. */
function isTruthy(v: string | undefined): boolean {
  if (!v) return false;
  return ["1", "true", "yes", "on"].includes(v.trim().toLowerCase());
}

function isSet(v: string | undefined): boolean {
  return typeof v === "string" && v.trim() !== "";
}

/**
 * What the guard decided, for callers that want to log rather than throw.
 *
 * - `ok`: authentication is on, or this is not a production build.
 * - `insecure-allowed`: auth is off and the operator opted in explicitly. Warn, loudly, forever.
 * - `refuse`: auth is off in production with no opt-in. `message` is the operator-facing text.
 */
export type AuthGuardVerdict =
  | { kind: "ok" }
  | { kind: "insecure-allowed"; message: string }
  | { kind: "refuse"; message: string };

function refusalMessage(devTenant: string): string {
  return `
================================================================================
ai-tally REFUSES TO START: ${DEV_TENANT_ENV} is set in a production build.

${DEV_TENANT_ENV} is the development escape hatch, and it disables authentication
COMPLETELY. With it set, the Clerk middleware becomes a pass-through, no
ClerkProvider is mounted, the tenant is pinned to

    ${devTenant}

instead of being resolved from a signed-in organization, and every visitor is
treated as an org admin who can mint, rotate and revoke API keys. Anyone who has
the URL sees every number in the system and can act on it.

There is no safe fallback here, so this process stops instead of serving.

You are in ONE OF TWO situations:

  1. You want a REAL, authenticated deployment. This is the normal case.

     Unset ${DEV_TENANT_ENV} (or set it to the empty string) and configure Clerk:

         NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY=pk_live_...
         CLERK_SECRET_KEY=sk_live_...

     The tenant then resolves from the signed-in Clerk organization, per
     web/lib/getTenant.ts.

     If you copied deploy/demo/, that export is the line to delete. The demo kit
     sets ${DEV_TENANT_ENV} deliberately and is NOT a template for a real
     instance: it serves synthetic data behind Caddy basic auth.

  2. You genuinely want an UNAUTHENTICATED instance, the way the public demo is.

     Then say so, explicitly, alongside ${DEV_TENANT_ENV}:

         ${ALLOW_INSECURE_ENV}=1

     Only do this behind access control you supply yourself, and only with
     synthetic data. deploy/demo/deploy.sh and deploy/demo/reseed.sh set it for
     you; you do not need to add it by hand there.

Checked: NODE_ENV=production, ${DEV_TENANT_ENV} set, ${ALLOW_INSECURE_ENV} not set.
================================================================================
`.trim();
}

function insecureWarning(devTenant: string): string {
  return `
================================================================================
ai-tally WARNING: serving with NO AUTHENTICATION.

${DEV_TENANT_ENV}=${devTenant} and ${ALLOW_INSECURE_ENV} is set, so this
production build starts with the dev escape hatch on: no sign-in, no Clerk org,
every visitor pinned to that tenant and treated as an org admin.

This is a deliberate configuration (the public demo runs this way behind basic
auth, with synthetic data). If you did not mean it, unset ${ALLOW_INSECURE_ENV}
and ${DEV_TENANT_ENV} and configure Clerk.
================================================================================
`.trim();
}

/**
 * Decide whether this process may serve, given its environment.
 *
 * Pure. `assertAuthConfig` is the one that actually stops the boot; this is separated out so tests
 * (and callers that want to log the allowed-but-insecure case) can inspect the decision.
 *
 * The production-BUILD phase is exempt. `next build` runs with NODE_ENV=production and CI builds
 * through the escape hatch on purpose (.github/workflows/ci.yml pins the same placeholder UUID the
 * vitest suite uses, because CI holds no Clerk keys). A build serves no requests; the artifact it
 * produces is checked when it boots, which is where the exposure would actually be.
 */
export function checkAuthConfig(env: AuthGuardEnv): AuthGuardVerdict {
  if (env.NEXT_PHASE === "phase-production-build") {
    return { kind: "ok" };
  }
  if (env.NODE_ENV !== "production") {
    return { kind: "ok" };
  }
  const devTenant = env.TALLY_DEV_TENANT;
  if (!isSet(devTenant)) {
    return { kind: "ok" };
  }
  const pinned = (devTenant as string).trim();
  if (isTruthy(env.TALLY_ALLOW_INSECURE_NO_AUTH)) {
    return { kind: "insecure-allowed", message: insecureWarning(pinned) };
  }
  return { kind: "refuse", message: refusalMessage(pinned) };
}

/**
 * Throw {@link InsecureAuthConfigError} when the escape hatch is active in a production build with
 * no explicit opt-in, and warn on stderr when it is active WITH one.
 *
 * Call sites are boot-time on purpose (see `instrumentation.ts`). A per-request check would be
 * skippable by any route that forgets to call it.
 */
export function assertAuthConfig(env: AuthGuardEnv = process.env as AuthGuardEnv): void {
  const verdict = checkAuthConfig(env);
  if (verdict.kind === "refuse") {
    throw new InsecureAuthConfigError(verdict.message);
  }
  if (verdict.kind === "insecure-allowed") {
    console.warn(verdict.message);
  }
}
